src/Compilers/Core/Portable/InternalUtilities/RoslynParallel.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+
+namespace Roslyn.Utilities
+{
+    internal static class RoslynParallel
+    {
+        internal static readonly ParallelOptions DefaultParallelOptions = new ParallelOptions();
+
+        /// <inheritdoc cref="Parallel.For(int, int, ParallelOptions, Action{int})"/>
+        public static ParallelLoopResult For(int fromInclusive, int toExclusive, Action<int> body, CancellationToken cancellationToken)
+        {
+            var parallelOptions = cancellationToken.CanBeCanceled
+                ? new ParallelOptions { CancellationToken = cancellationToken }
+                : DefaultParallelOptions;
+
+            return Parallel.For(fromInclusive, toExclusive, parallelOptions, errorHandlingBody);
+
+            // Local function
+            void errorHandlingBody(int i)
+            {
+                try
+                {
+                    body(i);
+                }
+                catch (Exception e) when (FatalError.ReportUnlessCanceled(e, cancellationToken))
+                {
+                    throw ExceptionUtilities.Unreachable;
+                }
+                catch (OperationCanceledException e) when (cancellationToken.IsCancellationRequested && e.CancellationToken != cancellationToken)
+                {
+                    // Parallel.For checks for a specific cancellation token, so make sure we throw with the
+                    // correct one.
```

Comment:
Curious: how did you discover this behavior? If there is a doc you read may be helpful to link to it here for future readers to find.

Reply:
I looked at referencesource.microsoft.com after observing a test failure.

Reply:
https://github.com/dotnet/runtime/blob/4c3d75927db24f745c3180cc3abedff3f37c6abd/src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.cs#L1217-L1221

Reply:
IMO this policy is slightly too aggressive. The API would be easier to use if it supported inner cancellation exceptions with _any_ cancellation token provided the one passed in the parallel options was also cancelled.